plugins/optimization-detective/detection.php

```diff
@@ -10,6 +10,54 @@
 	exit; // Exit if accessed directly.
 }
 
+/**
+ * Obtains the ID for a post related to this response so that page caches can be told to invalidate their cache.
+ *
+ * If the queried object for the response is a post, then that post's ID is used. Otherwise, it uses the ID of the first
+ * post in The Loop.
+ *
+ * When the queried object is a post (e.g. is_singular, is_posts_page, is_front_page w/ show_on_front=page), then this
+ * is the perfect match. A page caching plugin will be able to most reliably invalidate the cache for a URL via
+ * this ID if the relevant actions are triggered for the post (e.g. clean_post_cache, save_post, transition_post_status).
+ *
+ * Otherwise, if the response is an archive page or the front page where show_on_front=posts (i.e. is_home), then
+ * there is no singular post object that represents the URL. In this case, we obtain the first post in the main
+ * loop. By triggering the relevant actions for this post ID, page caches will have their best shot at invalidating
+ * the related URLs. Page caching plugins which leverage surrogate keys will be the most reliable here. Otherwise,
+ * caching plugins may just resort to automatically purging the cache for the homepage whenever any post is edited,
+ * which is better than nothing.
+ *
+ * There should not be any situation by default in which a page optimized with Optimization Detective does not have such
+ * a post available for cache purging. As seen in {@see od_can_optimize_response()}, when such a post ID is not
+ * available for cache purging then it returns false, as it also does in another case like if is_404().
+ *
+ * @since n.e.x.t
+ * @access private
+ *
+ * @return int|null Post ID or null if none found.
```

Comment:
Nit-pick, but why not stick with `int` and return `0` if none found? `0` can never be an ID anyway and is false-y too.

Reply:
True. But `null` seems to be more universally semantic to indicate nothing was found. If `null` then you can check that case with either `is_null()` or `!is_int()`. And `null` is also false-y.

Reply:
My counter-argument is that `int` is a single type, and the check can simply be made with `! $post_id` either way.

Reply:
Except not with our strict PHPStan rules in place 😄